### DISCUSSION TITLE
How to get two dot diff from Github API

### DISCUSSION THREAD

**Discussion by aTonyXiao** [community/community#24877] [Apps, API and Webhooks]
<p>Hello,</p>
<p>From this: <a href="https://help.github.com/articles/about-comparing-branches-in-pull-requests/#three-dot-and-two-dot-git-diff-comparisons" rel="nofollow noopener">https://help.github.com/articles/about-comparing-branches-in-pull-requests/#three-dot-and-two-dot-git-diff-comparisons</a></p>
<p>We are wondering if there is an api to get the two dot diff from two commits. </p>
<p>Thanks!</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hi <a class="mention" href="/atonyxiao">@atonyxiao</a>,</p>
<p>Thanks for being here! Check out the documentation on <a href="https://developer.github.com/v3/repos/commits/#compare-two-commits" rel="nofollow noopener">comparing commits across time</a>.</p>
<p>I hope this helps!</p>

**Comment by aTonyXiao** [comment]
<p>Hi <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a>,</p>
<p>Thanks for the reply. We read the documentation and find it cannot solve our problem.</p>
<p>Here is an example:</p>
<p>We want to compare two commits with sha </p>
<pre><code>4eff613a6c96579d11dad72e63200b74afc39433 and 2b58c5f5c142588b7b944f3b79a0371378da3db6
</code></pre>
<p>The API gives: <a href="https://api.github.com/repos/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433...2b58c5f5c142588b7b944f3b79a0371378da3db6" rel="nofollow noopener">https://api.github.com/repos/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433…2b58c5f5c142588b7b944f3b79a0371378da3db6</a></p>
<p>Which is the same as three dot diff on:</p>
<p><a href="https://github.com/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433...2b58c5f5c142588b7b944f3b79a0371378da3db6" rel="nofollow noopener">https://github.com/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433…2b58c5f5c142588b7b944f3b79a0371378da3db6</a></p>
<p>but we wonder if we can get this through API</p>
<p><a href="https://github.com/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433..2b58c5f5c142588b7b944f3b79a0371378da3db6" rel="nofollow noopener">https://github.com/python/mypy/compare/4eff613a6c96579d11dad72e63200b74afc39433…2b58c5f5c142588b7b944f3b79a0371378da3db6</a></p>
<p>Thanks!</p>

**Comment by AndreaGriffiths11** [comment, marked as answer]
<p>Hi <a class="mention" href="/atonyxiao">@atonyxiao</a>,</p>
<p>Thank you for taking the time to write back with more information. Unfortunately, there’s no way to do that with the GitHub REST API – you can’t pick which information will be returned. One option would be to pull each commit via the API or Git and then do a two-dot diff locally. </p>
<p>I hope this helps!</p>

**Comment by chrisdothtml** [comment]
<p><a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> is it accurate to say there’s no way with the API (or with a single request) to get the actual diff of a forced push (i.e. compare old head with new head)?</p>

**Comment by garyha** [comment]
<p>Surprised to see that go unanswered.</p>
<p>Edit: I had added an additional question which also went unanswered so I’ll answer it myself. I asked:</p>
<p>On the web, is there a way to diff this by version and all of .27 to .28?</p>
<p><a href="https://github.com/alpacahq/pylivetrader" rel="nofollow noopener">https://github.com/alpacahq/pylivetrader</a></p>
<p>Yes:</p>
<p><a href="https://github.com/alpacahq/pylivetrader/compare/v0.0.27..v0.0.28?expand=0" rel="nofollow noopener">https://github.com/alpacahq/pylivetrader/compare/v0.0.27…v0.0.28?expand=0</a> or</p>
<p><a href="https://github.com/alpacahq/pylivetrader/compare/v0.0.27...v0.0.28?expand=0" rel="nofollow noopener">https://github.com/alpacahq/pylivetrader/compare/v0.0.27…v0.0.28?expand=0</a></p>
<p>… or with expand=1</p>

**Comment by DenialAdams** [comment]
<p><span class="mention">@chrisdothtmldid</span> you come up with any alternatives to get the actual diff of a forced push? That was exactly my use case as well, and I can’t find a good way to do it (without an order of magnitude increase in API calls)</p>
<p>Thanks!</p>

**Comment by chrisdothtml** [comment]
<p>Not really. Depending on your use-case there are different approaches you can take. For example, if you need to know the net change of an entire pull request after a force push, you can just compare the base branch with the head branch.</p>
<p>I may be able to help more if you give a bit more info on why you need this as I spent quite a lot of time around this particular issue</p>
